src/rotation.jl

```diff
@@ -0,0 +1,286 @@
+    if iszero(θ)
+        out .= arr
+        return true 
+    end
+    # check for special cases where rotations are trivial
+    if (iseven(size(arr, 1)) && iseven(size(arr, 2)) && 
+        midpoint[1] ≈ size(arr, 1) ÷ 2 + 0.5 && midpoint[2] ≈ size(arr, 2) ÷ 2 + 0.5) ||
+        (isodd(size(arr, 1)) && isodd(size(arr, 2)) && 
+        (midpoint[1] == size(arr, 1) ÷ 2 + 1 && midpoint[1] == size(arr, 2) ÷ 2 + 1))
+        if θ ≈ π / 2 
+            if adjoint == false
+                out .= reverse(PermutedDimsArray(arr, (2, 1, 3, 4)), dims=(2,))
+            else
+                out .= reverse(PermutedDimsArray(arr, (2, 1, 3, 4)), dims=(1,))
+            end
+            return true
+        elseif θ ≈ π
+            out .= reverse(arr, dims=(1,2))
+            return true
+        elseif θ ≈ 3 / 2 * π
+            if adjoint == false
+                out .= reverse(PermutedDimsArray(arr, (2, 1, 3, 4)), dims=(1,))
+            else
+                out .= reverse(PermutedDimsArray(arr, (2, 1, 3, 4)), dims=(2,))
+            end
+            return true
+        end
+    end
+
+    return false
+end
+
+
+"""
+    imrotate(arr::AbstractArray{T, 4}, θ; method=:bilinear, midpoint=size(arr) .÷ 2 .+ 1)
+
+Rotates a matrix around the center pixel `midpoint`. `midpoint` is defined such that there
+is a real center pixel for even and odd values which is rotated around.
+The angle `θ` is interpreted in radians.
+
+The adjoint is defined with ChainRulesCore.jl. This method also runs with CUDA (and in principle all KernelAbstractions.jl supported backends).
+
+# Keywords
+* `method=:bilinear` for bilinear interpolation or `method=:nearest` for nearest neighbour
+* `midpoint=size(arr) .÷ 2 .+ 1` means there is a real center pixel around it is rotated.
+
+# Examples
+```julia-repl
+julia> arr = zeros((4,4,1,1)); arr[2,2,1,1] = 1;
+
+julia> arr
+4×4×1×1 Array{Float64, 4}:
+[:, :, 1, 1] =
+ 0.0  0.0  0.0  0.0
+ 0.0  1.0  0.0  0.0
+ 0.0  0.0  0.0  0.0
+ 0.0  0.0  0.0  0.0
+
+julia> NNlib.imrotate(arr, deg2rad(90)) # rotation around (3,3)
+4×4×1×1 Array{Float64, 4}:
+[:, :, 1, 1] =
+ 0.0  0.0  0.0  0.0
+ 0.0  0.0  0.0  1.0
+ 0.0  0.0  0.0  0.0
+ 0.0  0.0  0.0  0.0
+
+julia> NNlib.imrotate(arr, deg2rad(90), midpoint=(2,2))
+4×4×1×1 Array{Float64, 4}:
+[:, :, 1, 1] =
+ 0.0  0.0  0.0  0.0
+ 0.0  1.0  0.0  0.0
+ 0.0  0.0  0.0  0.0
+ 0.0  0.0  0.0  0.0
+
+julia> arr = zeros((3,3,1,1)); arr[1,2,1,1] = 1
+1
+
+julia> arr
+3×3×1×1 Array{Float64, 4}:
+[:, :, 1, 1] =
+ 0.0  1.0  0.0
+ 0.0  0.0  0.0
+ 0.0  0.0  0.0
+
+julia> NNlib.imrotate(arr, deg2rad(45))
+3×3×1×1 Array{Float64, 4}:
+[:, :, 1, 1] =
+ 0.0  0.207107  0.0
+ 0.0  0.0       0.207107
+ 0.0  0.0       0.0
+
+julia> NNlib.imrotate(arr, deg2rad(45), method=:nearest)
+3×3×1×1 Array{Float64, 4}:
+[:, :, 1, 1] =
+ 0.0  0.0  1.0
+ 0.0  0.0  0.0
+ 0.0  0.0  0.0
+```
+"""
+function imrotate(arr::AbstractArray{T, 4}, θ; method=:bilinear, midpoint=size(arr) .÷ 2 .+ 1) where T
```

Comment:
Does the midpoint always have to be the center pixel? Or does the algorithm support rotations around arbitrary points within the image? If arbitrary points are supported, I'd switch from midpoint to rotation_center. Also, is the default of sz/2+1 valid for all image sizes? If not I think it would be more secure to set the default to (0,0) and to check the sizes individually within the function body and set the default there to a value that is the center, no matter which size the image has.

Reply:
midpoint is meant to be the rotation_center, I change the name!

ImageTransformations does not allow for it.

There is also a difference to it. This routine always rotates around one pixel.
ImageTransformations rotates around the numerical center, which is in the case of even arrays at X.5

Reply:
In one case (90°) ImageTransformations rather looked a little sketchy